File: node/src/engine/commandSelector.ts
/**
 * Weighted command selection.
 *
 * Uses normalized cumulative weights, matching Java's inline CommandSelector
 * (BenchmarkEngine.java:520-545) and the Python engine. Selection intentionally
 * uses the platform RNG (`Math.random`), not the Java LCG: only *key* generation
 * must be cross-engine deterministic. Each worker owns its own selector, so
 * there is no shared state.
 */

import type { Command } from '../command/command.js';

export class CommandSelector {
  private readonly commands: Command[];
  private readonly cumulativeWeights: number[];

  constructor(commands: Command[]) {
    if (commands.length === 0) throw new Error('CommandSelector requires at least one command');
    this.commands = commands;
    this.cumulativeWeights = CommandSelector.buildCumulativeWeights(commands);
  }

  select(): Command {
    const r = Math.random();
    for (let i = 0; i < this.cumulativeWeights.length; i++) {
      if (r <= this.cumulativeWeights[i]!) return this.commands[i]!;
    }
    return this.commands[this.commands.length - 1]!;
  }

  private static buildCumulativeWeights(commands: Command[]): number[] {
    let totalWeight = commands.reduce((sum, command) => sum + command.weight, 0);
    if (totalWeight === 0) totalWeight = 1.0;

    const cumulative: number[] = [];
    let running = 0;
    for (const command of commands) {
      running += command.weight / totalWeight;
      cumulative.push(running);
    }
    return cumulative;
  }
}

File: go/internal/engine/commandselector.go
package engine

import (
	"math/rand"
	"sync"
	"time"

	"github.com/resp-bench/go/internal/command"
)

// CommandSelector picks a command according to its configured weight, using
// cumulative weights for O(n) selection. Each selector holds its own RNG; the
// choice RNG is intentionally NOT the Java-compatible one (command selection is
// not part of cross-engine key parity).
//
// A selector is shared across a connection's pipeline_depth worker goroutines,
// so Select is guarded by a mutex (math/rand.Rand is not concurrency-safe).
type CommandSelector struct {
	commands   []command.Command
	cumulative []float64
	mu         sync.Mutex
	rng        *rand.Rand
}

// NewCommandSelector builds a selector over the given commands.
func NewCommandSelector(cmds []command.Command) *CommandSelector {
	total := 0.0
	for _, c := range cmds {
		total += c.Weight()
	}
	if total == 0 {
		total = 1
	}
	cum := make([]float64, len(cmds))
	sum := 0.0
	for i, c := range cmds {
		sum += c.Weight() / total
		cum[i] = sum
	}
	return &CommandSelector{
		commands:   cmds,
		cumulative: cum,
		rng:        rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

// Select returns a command chosen by weight. Safe for concurrent callers.
func (s *CommandSelector) Select() command.Command {
	s.mu.Lock()
	r := s.rng.Float64()
	s.mu.Unlock()
	for i, threshold := range s.cumulative {
		if r <= threshold {
			return s.commands[i]
		}
	}
	return s.commands[len(s.commands)-1]
}
